EqSat/include/Luau/EGraph.h

```diff
@@ -0,0 +1,228 @@
+// This file is part of the Luau programming language and is licensed under MIT License; see LICENSE.txt for details
+#pragma once
+
+#include "Luau/Common.h"
+#include "Luau/Id.h"
+#include "Luau/Language.h"
+#include "Luau/UnionFind.h"
+#include "Luau/VecDeque.h"
+
+#include <optional>
+#include <unordered_map>
+#include <vector>
+
+namespace Luau::EqSat
+{
+
+template<typename L, typename N>
+class EGraph;
+
+template<typename L, typename N>
+struct Analysis final
+{
+    N analysis;
+
+    using D = typename N::Data;
+
+    template<typename T>
+    static D fnMake(const N& analysis, const EGraph<L, N>& egraph, const L& enode)
+    {
+        return analysis.make(egraph, *enode.template get<T>());
+    }
+
+    template<typename... Ts>
+    D make(const EGraph<L, N>& egraph, const Language<Ts...>& enode) const
+    {
+        using FnMake = D (*)(const N&, const EGraph<L, N>&, const L&);
+        static constexpr FnMake tableMake[sizeof...(Ts)] = {&fnMake<Ts>...};
+
+        return tableMake[enode.index()](analysis, egraph, enode);
+    }
+
+    void join(D& a, const D& b) const
+    {
+        return analysis.join(a, b);
+    }
+};
+
+/// Each e-class is a set of e-nodes representing equivalent terms from a given language,
+/// and an e-node is a function symbol paired with a list of children e-classes.
+template<typename L, typename D>
+struct EClass final
+{
+    Id id;
+    std::vector<L> nodes;
+    D data;
+    std::vector<std::pair<L, Id>> parents;
+};
+
+/// See <https://arxiv.org/pdf/2004.03082>.
+template<typename L, typename N>
+class EGraph final
```

Comment:
For consistency with the rest of the Luau analysis code, let's use `struct` everywhere.

Reply:
I can change it. I used `class`es in some spots to make stuff private by default.